### ISSUE TITLE
Restrict parent and child component hierarchy

### ISSUE THREAD

**Issue by Stamo-Gochev** [aspnet/AspNetCore#12302]
Tag helpers can restrict (to a certain extent) which tags can be their child tag helpers and vice versa using [RestrictChildren](https://github.com/aspnet/AspNetCore/blob/237a7a65d6432e0ac99537e73f745603cb9ad867/src/Razor/Razor/src/TagHelpers/RestrictChildrenAttribute.cs):
```
[HtmlTargetElement("parent-taghelper")]
[RestrictChildren("child-taghelper")]
public partial class ParentTagHelper : TagHelper { ... }
```
and also with [ParentTag](https://github.com/aspnet/AspNetCore/blob/ffec555118d6f45bc04c2de9108bcb93667d6b33/src/Razor/Razor/src/TagHelpers/HtmlTargetElementAttribute.cs#L84):
```
[HtmlTargetElement("child-taghelper", ParentTag="parent-taghelper" )]
public partial class ChildTagHelper : TagHelper { ... }
```
What will be the way to do that in blazor? Having the ability to restrict parent/child hierarchy will be very important for the IntelliSense as the users of components will not be able to define components in incorrect places.

### Describe the solution you'd like
How tag helpers are implemented to solve this issue (using strings) can be improved What I think will be better is to pass the name of the class that can be a parent of child in a way similar to:
```
[Child(MyNamespace.Components.ChildComponent)]
// multiple children
[Child(MyNamespace.Components.AnotherChildComponent)]
public class ParentComponent: ComponentBase
```
and the other way around:
```
[Parent(MyNamespace.Components.ParentComponent)]
// multiple parents
[Parent(MyNamespace.Components.AnotherParentComponent)]
public class ChildComponent : ComponentBase
```

The exact names of the attributes can be different. I am not sure whether they should be multiple attributes for a list of parents/children or a single one (e.g. `[Children(ChildComponent, AnotherChildComponent)]), but the general idea is to use concrete class names instead of strings.

You can also suggest other ways to solve this issue apart from using attributes. What are your thoughts? @SteveSandersonMS @nTay

### Additional context
The main problem

**Comment by mkArtakMSFT** [comment]
We've moved this issue to the Backlog milestone. This means that it is not going to happen for the coming release. We will reassess the backlog following the current release and consider this item at that time. However, keep in mind that there are many other high priority features with which it will be competing for resources.

**Comment by danroth27** [comment]
This is an ask from one of the major Blazor component vendors, so let's consider it for 5.0

**Comment by Stamo-Gochev** [comment]
Another option that can be discussed is to use an attribute for defining an "inner prop", e.g.:

```
public class ParentComponent: ComponentBase
{
    [ChildProperty]
    public ChildComponent ChildComponentInstance
    ...
}
```
By doing this, it is explicit that the `ChildComponent` should be visible from the IntelliSense when nested directly in the `ParentComponent` tag.

**Comment by jlchavez** [comment]
We have this as part of asp.net web forms controls, .net core razor pages with tag helpers, but it has taken too long for this to be added to blazor components. We need an easy way to construct complex components that can be linked so that it doesn't require much code to specify other nested objects as properties. We have templates but this hierarchical set of properties it's a must have, hopefully some day you add it.

**Comment by kyrvlasiuk** [comment]
Am I correct to assume if the feature is in the backlog at the start of work on a milestone, it will not be included in said milestone? Namely, no work is planned to address this feature in .NET 7.
Can anyone suggest a workaround to restrict content of RenderFragment or Component? 
Many Thanks

**Comment by danroth27** [comment]
> Am I correct to assume if the feature is in the backlog at the start of work on a milestone, it will not be included in said milestone? Namely, no work is planned to address this feature in .NET 7.

Correct. We think we've pulled the work out of the backlog at this point that we can get done for .NET 7. That's subject to change, of course, based on community feedback, availability of resources, etc.

**Comment by eyalalonn** [comment]
I've added a feature request related to this issue #42126.

**Comment by viniciusmiguel** [comment]
This is dragging for years, how long will take to get it done?
